---
title: "CSU Canine Leukemia"
author: "Jenny Smith"
date: "3/12/21"
output: html_document
---

# Set-up 

```{r setup, cache = FALSE, include = FALSE}
require(knitr)
knitr::opts_knit$set(root.dir = file.path(PROJHOME,"2021.03.08_CSU_Canine_AML"))
```

```{r}
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),
                      tidy=TRUE,
                      fig.align='center',
                      fig.width = 10, fig.height = 10)
node=Sys.info()[["nodename"]]
if(!grepl("local", node)){
  print(node)
  options(bitmapType = 'cairo')
  grDevices::X11.options(type='cairo')
}

options(stringsAsFactors = FALSE)
table = function (..., useNA = 'ifany') base::table(..., useNA = useNA)
```

```{r message = FALSE, warning=FALSE}
library(stringr)
library(magrittr)

library(ggplot2)
library(gridExtra)

library(dplyr)
library(tidyr)
library(tibble)


library(DeGSEA)

getwd()
```


# ClinData

```{r}
canine_info <- read.csv("CSU_Canine_Leukemia_sample_manifest_v2.csv")

head(canine_info)
dim(canine_info)
```

```{r}
# cts_dir <- file.path(SCRATCH,"jlsmith3/CSU_Canine_AML/counts")
# files <- dir(cts_dir, full.names = TRUE,  pattern=".tab")
# IDmap <- str_split_fixed(files, pattern = "_", n=10)[,c(4,5)] %>%
#   as.data.frame() %>%
#   separate(V1, into=c("path","Library"), sep="-") %>%
#   select(Library, rnaseq_identifier=V2) %>%
#   left_join(., rnaseq_identifier_info, by="rnaseq_identifier")

 
# IDmap
# write.csv(IDmap, "CSU_Canine_Leukemia_rnaseq_identifier_manifest_v2.csv", row.names = FALSE)
```

# Gene Annotatations 

```{r}
AML_partners <- read.csv("References/TARGET_AML_refSeq_ensembl_transcript_IDmap.csv")

# head(AML_partners)
# dim(AML_partners)
```

```{r}
orthologs <- read.csv("References/canFam_Human_homologs_orthologs.csv") %>% 
  janitor::clean_names()


# head(orthologs)
# dim(orthologs)
```

```{r}
common_aml_fusions <- orthologs %>% 
  select(1:2,gene_name,human_gene_name,
         human_gene_order_conservation_score,
         human_orthology_confidence_0_low_1_high,
         human_whole_genome_alignment_coverage,
         everything()) %>% 
  filter(gene_name %in% AML_partners$external_gene_name)

# common_aml_fusions
```


# BCCA 

```{r}
outpath=file.path(SCRATCH,"jlsmith3/CSU_Canine_AML/BCCA_fusion")
fusion_res <- dir(outpath, pattern="^sense_fusion.tsv",
                  recursive=TRUE, full.names = TRUE) %>% 
  grep("odd",., invert=TRUE, value=TRUE)

head(fusion_res)
length(fusion_res)
```

```{r}
fusions <- purrr::map_dfr(.x=fusion_res, .f=function(x){
  read.delim(x, sep="\t", header=TRUE) %>% 
    mutate_at(vars(matches("size|id")), ~as.character(.)) %>% 
    mutate(Library=gsub("^.+fusion\\/([K][0-9]+)\\/.+","\\1", x))
}) %>% 
  left_join(., select(canine_info, Library, rnaseq_identifier),
            by="Library") %>% 
   select(Library,rnaseq_identifier, everything())


head(fusions)
dim(fusions) #4681   36


# length(unique(fusions$Library)) #133
# write.csv(fusions, "Fusions/CSU_Canine_AML_BCCA_TransAbyss_Fusion_original.csv", row.names = FALSE)
```

```{r}
# fusions %>% 
#   filter(grepl("KMT2A|NUP98|CBFB|RUNX1", X.FusionName))
regex <- paste(common_aml_fusions$gene_name, collapse = "|")

# fusions %>%
#   filter(grepl(regex, genes)) %>%
#   arrange(genes) %>% 
#   select(genes, everything())
```

The number of fusions appears to be very, very low. I will need to ask which paramters were used for TransAbyss-fusion by BCCA.


# Reformat and Clean Fusion Cols

```{r}
fusions <- read.csv("Fusions/CSU_Canine_AML_BCCA_TransAbyss_Fusion_original.csv")
dim(fusions)
```

```{r}
sense.fmt <- fusions %>% 
    arrange(rnaseq_identifier) %>%
    
    mutate(Fusion.Detected.TransAbyss=ifelse(is.na(genes), "NoneDetected","Detected")) %>% 
    mutate_at(vars(c("X5.gene","X3.gene")), ~gsub("-","\\.", .)) %>%  #replace dashes with a dot. 

    mutate(b1=str_split_fixed(breakpoint,"\\|", n=2)[,1],
           b2=str_split_fixed(breakpoint,"\\|", n=2)[,2]) %>%
    mutate(b1=paste(b1,str_split_fixed(genes,",", n=2)[,1], sep="_"),
           b2=paste(b2,str_split_fixed(genes,",", n=2)[,2], sep="_")) %>% 
  
    mutate(t1=str_split_fixed(transcripts,",", n=2)[,1],
           t2=str_split_fixed(transcripts,",", n=2)[,2]) %>%
    mutate(t1=paste(t1,str_split_fixed(genes,",", n=2)[,1], sep="_"),
           t2=paste(t2,str_split_fixed(genes,",", n=2)[,2], sep="_")) %>% 
    
    #Needed to rearrange breakpoints to be in same orientation at X5.gene, X3. gene
    #Original TransAbyss data has breakpoints sorted by genomic coordinate, not which partenr was 5' or 3'
    rowwise() %>%
    
    mutate(X5.breakpoint = case_when(
              str_detect(b1, X5.gene) ~ str_split_fixed(b1, "_",n=2)[,1],
              str_detect(b2, X5.gene) ~ str_split_fixed(b2, "_",n=2)[,1]), 
           X3.breakpoint = case_when(
              str_detect(b1, X3.gene) ~ str_split_fixed(b1, "_",n=2)[,1],
              str_detect(b2, X3.gene) ~ str_split_fixed(b2, "_",n=2)[,1]), 
           
           X5.transcript = case_when(
              str_detect(t1, X5.gene) ~ str_split_fixed(t1, "_",n=2)[,1],
              str_detect(t2, X5.gene) ~ str_split_fixed(t2, "_",n=2)[,1]),
           X3.transcript = case_when(
              str_detect(t1, X3.gene) ~ str_split_fixed(t1, "_",n=2)[,1],
              str_detect(t2, X3.gene) ~ str_split_fixed(t2, "_",n=2)[,1]),) %>%

    ungroup() %>% 
    
    #update the genes that are lncRNAs to the transcript IDs
    mutate_at(vars(X5.gene),~ifelse(grepl("none|cmpl", .), X5.transcript, .)) %>% 
    mutate_at(vars(X3.gene),~ifelse(grepl("none|cmpl", .),  X3.transcript, .)) %>% 
  
    #create standard fusion name columns
    mutate(Fusion=ifelse(! is.na(genes), paste(X5.gene, X3.gene, sep="-"), NA),
           Type=ifelse(! is.na(genes), ifelse(grepl("^[0-9]", size),   "intrachromosomal", "interchromosomal"), NA)) %>%
    
    select(rnaseq_identifier,Library,
           Fusion,breakpoint,
           X5.breakpoint,X3.breakpoint,
           everything(),
           -c(b1:t2)) %>%
    # mutate_at(vars(3:NCOL(.)), ~as.character(.)) %>%
    # mutate_at(vars(3:NCOL(.)), ~ifelse(grepl("NoneDetected", Fusion.Detected.TransAbyss),  "NoneDetected", .)) %>%
    ungroup()
    
```

```{r}
head(sense.fmt)

# length(unique(sense.fmt$rnaseq_identifier)) #133 rnaseq_identifiers
# dim(sense.fmt)  #4681   42

# write.csv(sense.fmt, "Fusions/CSU_Canine_AML_BCCA_TransAbyss_Fusion_formatted_10.28.2021.csv", row.names = FALSE)
```


# Filter out Fusions found in NBMs 

```{r}
normals <- canine_info %>% 
  filter(grepl("Normal", tentative_dx))

fusions_normal <- sense.fmt %>% 
  filter(rnaseq_identifier %in% normals$rnaseq_identifier)

dim(fusions_normal) #532  41
# length(unique(fusions_normal$rnaseq_identifier))
```

```{r warning=FALSE}
sense.fmt.filter <- sense.fmt %>%
  mutate_at(vars(3:NCOL(.)),~gsub("/", "-", .)) %>% #to address the slash at "TNRC6Cexon3-septin9/TNRC6Cfusionexon1""
  mutate_at(vars(3:NCOL(.)),~gsub(" ", "", .)) %>% #to address the slash at "TNRC6Cexon3-septin9/TNRC6Cfusionexon1""
  mutate_at(vars(3:NCOL(.)), ~gsub("@","", .)) %>%
  
  mutate(fusion_in_controls=breakpoint %in% fusions_normal$breakpoint) %>% 
  
  group_by(rnaseq_identifier) %>% 
  mutate(all_fail=case_when(
    sum(fusion_in_controls) == n() ~ TRUE,
    TRUE ~ FALSE)) %>% 
  ungroup() %>% 
  
  mutate(across(.cols=c(Fusion:Type), .fns = ~case_when(
    is.character(.x) & all_fail ~ "NonePassedFilter", 
    is.numeric(.x) & all_fail  ~ NA_character_,
    TRUE ~ as.character(.)))) %>% 
  distinct() %>% 
  
  filter(!fusion_in_controls | Fusion=="NonePassedFilter") %>%
  
  # select(rnaseq_identifier,Library, everything(), 
  #        -probe, #hmmm so in retrospect, this probe == fusion junc sequence. so its very useful. I need to merge these back in at the end. by=c(rnaseq_identifier, Breakpoint)
  #        -AllFail,-FailedFilterFusions,
  #        -alignment_params, -dbsnp, -dgv, -id,
  #        -senses, -strands, -genomic_regions,
  #        -contig, -contig_regions, -contig_size) %>%
  # ungroup() %>%
  distinct() #remove the duplicate entries with different senses/strands, but everyhting else is identical. Note: flanking_pairs are often the ONLY difference between two rows.


head(sense.fmt.filter)
dim(sense.fmt.filter) #2254   43
# length(unique(sense.fmt.filter$rnaseq_identifier))
```

```{r}
# write.csv(sense.fmt.filter,
#           "Fusions/CSU_Canine_AML_BCCA_TransAbyss_Fusion_reformatted_FilteredForNormals_10.28.21.csv",
#           row.names = FALSE)

```


# Define the Primary Fusions/Breakpoints per patient 

```{r}
#Helper Functions for Tidy Cleaning
fusionCategory <- function(geneA,geneB){
  fus <- c(geneA,geneB)
  fus <- paste(fus[order(fus)], collapse = "-") #order alphabetically, so that the category includes  reciprocal fusions
}

collapseFusions <- function(breakpoint){
  paste(unique(breakpoint), collapse = "; ")
}

fillNA <- function(x,USI){
  
  USI <- unique(USI)
  x <- x %>% .[!is.na(.)] %>% unique()
  
  if(length(x) < 1){
    x <- ""
  }
  
  # print(x)
  return(x)
}
```


```{r}
getPrimaryFusion <- function(USI,flanking_pairs,sum.break.reads,spanning_reads,in_frame){
  # breakpoint, flanking_pairs,breakpoint_pairs,spanning_reads,frame are all columns from TransAbyss
  # brk.reads <- max(as.numeric(str_split_fixed(breakpoint_pairs,",", n=2)))
  
  if (all(is.na(sum.break.reads))){
    res <- NA
    return(res)
  }
  
  resolveTies <- function(df){
        df <- df %>%
            group_by(primary) %>%
            mutate(rank=rank(flanking_pairs, ties.method = "first")) %>% #if still a tie in flanking pairs counts, choose the first. 
            mutate(rank=ifelse(primary=="alternate", 0, rank)) %>%
            rowwise() %>%
            mutate(reason=sum(reason,rank)) %>%
            ungroup() %>% 
            mutate(primary=ifelse(reason == max(reason), "primary", "alternate"))
        return(df)
    }
  
  #Create boolean vectors for each type of evidence supporting the "primary" fusion. 
  USI <- unique(USI)
  frames <- ifelse(grepl("in", in_frame),4,0) #wieghted so that in frame is always the primary regardless of having fewer counts.
  brk.reads <-  sum.break.reads == max(sum.break.reads)
  spanning.reads <- spanning_reads == max(spanning_reads)
  flanking.reads <- flanking_pairs == max(flanking_pairs)
  
  temp <- list("frames"=frames,"brk.reads"=brk.reads, 
               "spanning.reads"=spanning.reads, "flanking.reads"=flanking.reads)
  
  #Create a temporary data frame. 
  temp <- bind_cols(temp) %>% 
    bind_cols(.,list("flanking_pairs"=flanking_pairs, "sum.break.reads"=sum.break.reads,
                     "spanning_reads"=spanning_reads,"in_frame"=in_frame))
  
  #Will prioritize the in-frame fusions (weighted), then select that one with the highest reads
    temp <- temp %>%
        rowwise() %>%
        mutate(reason=sum(frames,brk.reads,spanning.reads,flanking.reads)) %>%
        ungroup() %>%
        mutate(primary=ifelse(reason == max(reason), "primary", "alternate"))

    # break ties if there are multiple fusions with same #s of supporting reads.
    # for those with highest evidence ("reason"), choose the one with the higher flanking reads
    if (sum(temp$primary == "primary") > 1){
        temp <- resolveTies(temp)
        if (sum(temp$primary == "primary") >1 ){
          # last check for ties, will break dplyr.
          return(temp)
        }
    }

  res <- temp$primary
  
  return(res)
}
```

```{r warning=FALSE}
sense.fmt.primary <- sense.fmt.filter %>% 
  select(everything(), In_frame=frame) %>% #to avoid confusion with an r function called frame()
  
  rowwise() %>%
  #Fusion category is to encompass both reciprocal fusions under 1 type/name. 
  mutate(Fusion.Category=fusionCategory(X5.gene,X3.gene)) %>% 
  #using sum of the breakpoint pairs. 
  mutate(sum.break.reads=sum(as.numeric(str_split_fixed(breakpoint_pairs,",", n=2)))) %>% 
  ungroup() %>%

  group_by(rnaseq_identifier,Fusion.Category) %>%
  mutate(Primary.Fusion=getPrimaryFusion(USI=Library,flanking_pairs = flanking_pairs, 
                                         sum.break.reads = sum.break.reads, 
                                         spanning_reads = spanning_reads, in_frame = In_frame)) %>%
  arrange(rnaseq_identifier,Fusion.Category) %>% 
  
  group_by(rnaseq_identifier,Fusion.Category,Primary.Fusion) %>%
  mutate(Alternate.Breakpoints=ifelse(Primary.Fusion=="alternate", 
                                      collapseFusions(breakpoint = breakpoint), NA)) %>%
  ungroup() %>%
  
  group_by(rnaseq_identifier,Fusion.Category) %>%
  mutate(Alternate.Breakpoints=fillNA(Alternate.Breakpoints, USI=Library)) %>% 
  mutate(Alternate.Breakpoints=ifelse(breakpoint==Alternate.Breakpoints,"",Alternate.Breakpoints)) %>%
  ungroup() %>% 
  
  filter(grepl("primary", Primary.Fusion) | grepl("NonePassedFilter|NoneDetected", Fusion)) %>%
  select(rnaseq_identifier,Library,
         Fusion.Category,Fusion, X5.breakpoint, X3.breakpoint,
         Alternate.Breakpoints, everything(),-Primary.Fusion)


dim(sense.fmt.primary) #1920   48
length(unique(sense.fmt.primary$rnaseq_identifier)) #133
head(sense.fmt.primary)
```

```{r}
# write.csv(sense.fmt.primary, "Fusions/CSU_Canine_AML_BCCA_TransAbyss_Fusion_reformatted_FilteredForNormals_PrimaryFusions_10.28.2021.csv", row.names = FALSE)
```


#Session Information

```{r}
sessionInfo()
```

